x/evm/keeper/state_transition.go

```diff
@@ -345,7 +313,25 @@ func (k *Keeper) ApplyMessageWithConfig(msg core.Message, tracer vm.Tracer, comm
 	if contractCreation {
 		ret, _, leftoverGas, vmErr = evm.Create(sender, msg.Data(), leftoverGas, msg.Value())
 	} else {
+		// Increment the nonce for the next transaction
+		k.SetNonce(msg.From(), k.GetNonce(sender.Address())+1)
+
+		// When hooks execution failed, we need to revert tx execution side effects, but not nonce increasement,
+		// but we can't do that for contract creation, because we can't separate out the nonce increase there.
```

Comment:
can we revert and then re-increment the nonce?

Reply:
yeah, maybe decrease nonce after contract creation succeeds, there will be less code change.

Reply:
have redo the PR, much simpler now.